core/optaplanner-constraint-streams/src/main/java/org/optaplanner/constraint/streams/drools/common/AbstractPatternVariable.java

```diff
@@ -0,0 +1,250 @@
+            List<ViewItem<?>> prerequisiteExpressions, List<ViewItem<?>> dependentExpressions) {
+        this.primaryVariable = aVariable;
+        this.patternSupplier = patternSupplier;
+        this.prerequisiteExpressions = prerequisiteExpressions;
+        this.dependentExpressions = dependentExpressions;
+    }
+
+    protected AbstractPatternVariable(AbstractPatternVariable<?, PatternVar_, ?> patternCreator,
+            Variable<A> boundVariable) {
+        this.primaryVariable = boundVariable;
+        this.patternSupplier = patternCreator.getPatternSupplier();
+        this.prerequisiteExpressions = patternCreator.getPrerequisiteExpressions();
+        this.dependentExpressions = patternCreator.getDependentExpressions();
+    }
+
+    protected AbstractPatternVariable(AbstractPatternVariable<A, PatternVar_, ?> patternCreator,
+            UnaryOperator<PatternDSL.PatternDef<PatternVar_>> patternMutator) {
+        this.primaryVariable = patternCreator.primaryVariable;
+        this.patternSupplier = () -> patternMutator.apply(patternCreator.patternSupplier.get());
+        this.prerequisiteExpressions = patternCreator.prerequisiteExpressions;
+        this.dependentExpressions = patternCreator.dependentExpressions;
+    }
+
+    protected AbstractPatternVariable(AbstractPatternVariable<A, PatternVar_, ?> patternCreator,
+            ViewItem<?> dependentExpression) {
+        this.primaryVariable = patternCreator.primaryVariable;
+        this.patternSupplier = patternCreator.patternSupplier;
+        this.prerequisiteExpressions = patternCreator.prerequisiteExpressions;
+        this.dependentExpressions = Stream.concat(patternCreator.dependentExpressions.stream(), Stream.of(dependentExpression))
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public Variable<A> getPrimaryVariable() {
+        return primaryVariable;
+    }
+
+    public Supplier<PatternDSL.PatternDef<PatternVar_>> getPatternSupplier() {
+        return patternSupplier;
+    }
+
+    @Override
+    public List<ViewItem<?>> getPrerequisiteExpressions() {
+        return prerequisiteExpressions;
+    }
+
+    @Override
+    public List<ViewItem<?>> getDependentExpressions() {
+        return dependentExpressions;
+    }
+
+    /**
+     * Variable values can be either read directly from the pattern variable (see {@link DirectPatternVariable}
+     * or indirectly by applying a mapping function to it (see {@link IndirectPatternVariable}.
+     * This method abstracts this behavior, so that the surrounding code may be shared between both implementations.
+     *
+     * @param patternVar never null, pattern variable to extract the value from
+     * @return value of the variable
+     */
+    protected abstract A extract(PatternVar_ patternVar);
+
+    protected abstract Child_ create(UnaryOperator<PatternDSL.PatternDef<PatternVar_>> patternMutator);
+
+    protected abstract Child_ create(ViewItem<?> dependentExpression);
```

Comment:
The extending classes are now required to implement this.
All the other code in this class has been moved from the extending classes, without modification.